### PR TITLE
Adding extra rules for Java, C#, PHP, Python, Node.js, Apache, nginx …

### PR DESCRIPTION
**Pull request by thistehneisen**
…and MongoDB

### CODE REVIEW

**Comment by augustd** [comment]
Thanks! Do you have any real world examples of responses that would trigger these rules? 

This is needed in order to update the unit tests: https://github.com/augustd/burp-suite-error-message-checks/tree/master/src/test/resources/burp

**Comment by thistehneisen** [comment]
Hi there, @augustd!

Here you go (same order):
```
java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
java.lang.IllegalArgumentException: Argument is not an array
java.lang.IllegalStateException: Could not execute method for android:onClick
java.io.FileNotFoundException: File (filename.txt) not found in the specified path
System.NullReferenceException: Object reference not set to an instance of an object.
System.ArgumentNullException: Value cannot be null. (Parameter 'arg')
System.IO.FileNotFoundException: Could not find file 'C:\myfolder\myfile.txt'.
System.InvalidOperationException: Sequence contains no elements
Notice: Undefined index: mykey in C:\xampp\htdocs\test\index.php on line 15
Notice: Undefined variable: varName in C:\xampp\htdocs\test\index.php on line 15
Notice: Undefined offset: 5 in C:\xampp\htdocs\test\index.php on line 15
NameError: name 'undefinedVar' is not defined
ImportError: No module named 'nonexistentmodule'
IndentationError: expected an indented block
ReferenceError: nonExistentFunction is not defined
TypeError: Cannot read property 'prop' of undefined
UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: This is an error text
AH00526: Syntax error on line 20 of /etc/apache2/apache2.conf: Invalid command 'undefCommand', perhaps misspelled or defined by a module not included in the server configuration
mod_evasive: client 192.168.1.101 has reached or exceeded the 'maxconnsperip' limit.
client intended to address ipv6 backend without ipv6 on socket
could not build optimal proxy_headers_hash, you should increase either proxy_headers_hash_max_size: 512 or proxy_headers_hash_bucket_size: 64; ignoring proxy_headers_hash_bucket_size
E QUERY [thread1] SyntaxError: invalid object initializer
uncaught exception: out of memory
```

File: src/main/resources/burp/match-rules.tab [hunk 1]
```diff
@@ -1,73 +1,89 @@
...
+Oracle.*Driver]	0	Oracle	Low	Certain	0
+Warning.*\Woci_.*	0	Oracle	Low	Certain	0
+Warning.*\Wora_.*	0	Oracle	Low	Certain	0
+Warning: oci_parse()	0	Oracle	Low	Certain	0
+at (\/[A-Za-z0-9\.]+)*\.pm line [0-9]+	0	Perl	Low	Certain
+\.php on line [0-9]+	0	PHP	Low	Certain	4
+\.php</b> on line <b>[0-9]+	0	PHP	Low	Certain
+Fatal error:	0	PHP	Low	Certain	2
+\.php:[0-9]+	0	PHP	Low	Certain
+Undefined (index|variable|offset):	0	PHP	Low	Certain	3
```
Comment (augustd) [inline src/main/resources/burp/match-rules.tab:72]: Combined 3 rules into one for better performance.

File: src/main/resources/burp/match-rules.tab [hunk 2]
```diff
@@ -1,73 +1,89 @@
...
 ((dn|dc|cn|ou|uid|o|c)=[\w\d]*,\s?){2,}	0	LDAP	Low	Firm
+DB Error:	0	Maria	Low	Certain	0
 \[(ODBC SQL Server Driver|SQL Server|ODBC Driver Manager)\]	0	Microsoft SQL Server	Low	Certain
+Unclosed quotation mark	0	Microsoft SQL Server	Low	Certain	1
+warning.*mssql_.*	0	Microsoft SQL Server	Low	Certain	0
+Driver.* SQL[-_]*Server	0	Microsoft SQL Server	Low	Certain	0
+(\W|\A)SQL Server.*Driver	0	Microsoft SQL Server	Low	Certain	0
+Conversion failed when converting the	0	Microsoft SQL Server	Low	Certain	0
 Cannot initialize the data source object of OLE DB provider "[\w]*" for linked server "[\w]*"	0	Microsoft SQL Server	Low	Certain
+E QUERY\s+\[thread1\] SyntaxError:	0	MongoDB	Low	Certain	2
```
Comment (augustd) [inline src/main/resources/burp/match-rules.tab:45]: Noticed from searching that the number of spaces in the response can vary.

File: src/main/resources/burp/match-rules.tab [hunk 3]
```diff
@@ -1,73 +1,89 @@
...
+Exception.*Informix	0	Informix	Low	Certain	0
+Warning.*ingre_	0	Ingres DB	Low	Certain	0
+Ingres SQLSTATE	0	Ingres DB	Low	Certain	0
+Ingres\W.*Driver	0	Ingres DB	Low	Certain	0
+HSQLDB	0	Ingres DB	Low	Certain	0
 \.java:[0-9]+	0	Java	Low	Certain
 \.java\((Inlined )?Compiled Code\)	0	Java	Low	Certain	2
 \.invoke\(Unknown Source\)	0	Java	Low	Certain
 nested exception is	0	Java	Low	Firm
+java\.lang\.([A-Za-z0-9_]*)Exception	0	Java	Medium	Firm	5
```
Comment (augustd) [inline src/main/resources/burp/match-rules.tab:32]: Combined several individual exception rules into a single wildcard for performance.

File: src/test/java/burp/RegexTest.java [hunk 4]
```diff
@@ -66,39 +64,77 @@ public void testMatchRules() {
...
         
         for (MatchRule rule : matchRules) {
-			System.out.print("Testing rule: " + rule.getPattern());
+	        int expectedMatches = (rule.getExpectedMatches() != null) ? rule.getExpectedMatches() : 1 ;
+			expectedCount += expectedMatches;
+			System.out.print("Testing rule: '" + rule.getPattern() + "' expected: " + expectedMatches);
 			Matcher matcher = rule.getPattern().matcher(testResponse);
-            long startTime = System.currentTimeMillis();
-            int expectedMatches = (rule.getExpectedMatches() != null) ? rule.getExpectedMatches() : 1 ;
+            long startTime = System.nanoTime();
```
Comment (augustd) [inline src/test/java/burp/RegexTest.java:74]: Changed to nanoseconds to more accurately measure performance.

**Review by augustd — APPROVED** [review]
Thank you for your submission! Keep them coming.

**Pull request merged**